PT-152575233 Add all block transactions' fees to miner's account

zp-sd opened the pull request:
* Remove trees() argument from aec_tx:new/2
* [Finishes #152575233]

zp-sd commented on rebar.config:
```diff
@@ -25,7 +25,7 @@
 
 {plugins,
-  [{tx_xform, {git, "https://github.com/aeternity/tx_xform.git", {branch, "master"}}}]}.
+  [{tx_xform, {git, "https://github.com/zp-sd/tx_xform.git", {branch, "change-arity-of-aec_tx_new"}}}]}.
```
See https://github.com/aeternity/tx_xform/pull/3.

lucafavatella replied:
@zp-sd I would suggest you:
* Push branch change-arity-of-aec_tx_new in https://github.com/aeternity/tx_xform repo
* Make this rebar.config line refer to https://github.com/aeternity/tx_xform.git specific commit (rather than branch) i.e. `{ref, "fa96672"}` (referring to specific commit rather than master helps dealing with concurrent PRs as in this case)

zp-sd replied:
This depends on https://github.com/aeternity/epoch/pull/325 and https://github.com/aeternity/tx_xform/pull/3.

sennui commented on apps/aecore/src/txs/aec_tx.erl:
```diff
@@ -78,3 +106,26 @@ check_single(Tx, Trees, Height) ->
+
+    {ok, AccountsTrees} = aec_accounts:put(Account, AccountsTrees0),
+    Trees = aec_trees:set_accounts(Trees0, AccountsTrees),
+    Trees.
```
@uwiger @gorillainduction mind that there is no transaction representing update to miners account by sum of all fees

lucafavatella replied:
I see that coinbase tx has the responsibility of identifying the account of the miner, and contains no amont.
@zp-sd Could you please add comment in [`aec_coinbase_tx:process`](https://github.com/aeternity/epoch/pull/322/files#diff-933a22fcb74313128547fe0436956dd9R37) mentioning that the earning of the total of the non-coinbase tx fees is not computed inside the coinbase transaction but rather at the end of the application of all transactions in the block?

zp-sd replied:
@lucafavatella Coinbase tx still adds `block_mine_reward` to miner's account, which is driven by "governance".

sennui commented on apps/aecore/src/txs/aec_tx.erl:
```diff
@@ -40,29 +40,57 @@ fee(#spend_tx{fee = F}) ->
+verify_and_extract_txs([], Txs) ->
+    lists:reverse(Txs);
+verify_and_extract_txs([SignedTx | Rest], Txs) ->
+    case aec_tx_sign:verify(SignedTx) of
```
this is mock'ed. @uwiger implements the verify API

lucafavatella commented on apps/aecore/test/txs/aec_tx_tests.erl:
```diff
@@ -0,0 +1,96 @@
+
+%% Internals
+
+init_keys() ->
```
I see these test functions are now common and placed in common module https://github.com/aeternity/epoch/blob/3ddf1b87250adc5e64289d7d5cd7b4b25fa94335/apps/aecore/test/aec_test_utils.erl#L17-L18

(No need to change this PR only for this though.)

lucafavatella commented on apps/aecore/src/aec_mining.erl:
```diff
@@ -67,10 +63,10 @@ mine(Block, Attempts, InitialNonce, MaxNonce) ->
+get_txs_to_mine() ->
     {ok, Txs0} = aec_tx_pool:peek(aec_governance:max_txs_in_block() - 1),
-    case create_coinbase_tx(Trees) of
+    case create_coinbase_tx() of
```
I see that other transactions in the block are not passed to creation of the signed coinbase tx, as the amount is not part of the coinbase tx. Ok.

zp-sd replied:
Amount is part of coinbase tx. We simply don't need state trees to create coinbase tx. Amount is taken from simplified governance atm.

lucafavatella commented on apps/aecore/src/txs/aec_tx.erl:
```diff
@@ -40,29 +40,57 @@ fee(#spend_tx{fee = F}) ->
     F.
 
 -spec apply_signed(list(signed_tx()), trees(), non_neg_integer()) ->
-                          {ok, trees()} | {error, term()}.
```
This function `aec_tx:apply_signed` changed its API to always succeed by silently discarding invalid txs. I see this same function is used by the module [building the state trees in the chain](https://github.com/aeternity/epoch/blob/b0950fb9b0ddef9653a720486627c09ffff17c8c/apps/aecore/src/aec_chain_state.erl#L530) that shall see application of all txs succeed.

Please separate function in two:
* Once expecting all txs to be valid, and returning error otherwise (I envisage this to be used by (1) module processing block received from network and (2) module building state trees in chain);
* One silently discarding transactions (for mining - as proposed in this PR).

zp-sd replied:
I believe that the fact that txs got applied on state tree without failures should be verified by checking hash of state tree (`#block.root_hash` field) when absorbing a block. If a tx cannot be applied, hash simply will not match. And checking `root_hash` is the last thing to be implemented for `PT-#152481000 Validate received block`. I'm not sure if separating this function brings much value then, let me think about it.

lucafavatella commented on apps/aecore/src/txs/aec_tx.erl:
```diff
@@ -40,29 +40,57 @@ fee(#spend_tx{fee = F}) ->
-    end.
+                          {ok, trees()}.
+apply_signed(SignedTxs, Trees0, Height) ->
+    Txs = verify_and_extract_txs(SignedTxs),
```
As function `verify_and_extract_txs` silently discards txs, the coinbase tx may be silently discarded too!
Please make a proposal where in the code there is the hardcoded expectation that the first tx among the txs being included in the block to be mined is the coinbase tx (so that the reader does not need to assume so with the risk that block w/o coinbase transaction is mined and then rejected by other nodes in network).

zp-sd replied:
There is a check on coinbase tx coint in https://github.com/aeternity/epoch/blob/master/apps/aecore/src/aec_blocks.erl#L246 (it is executed when accepting a block from the network), so we may explicitly expect 1 coinbase tx there, wdyt?

lucafavatella replied:
> There is a check on coinbase tx coint in https://github.com/aeternity/epoch/blob/master/apps/aecore/src/aec_blocks.erl#L246 (it is executed when accepting a block from the network), so we may explicitly expect 1 coinbase tx there, wdyt?

If that function is not called on genesis (I did not check, seems out of scope for this PR), check of equal to 1 is ok.

Though wrt mining, I did not see validation of coinbase tx on block candidate (please let me know if I missed it).

zp-sd commented:
@lucafavatella Regarding `Review impact of silently discarding txs in apply_signed`: I think this is what needs to be done. Imagine the situation that we have a signed tx in our mempool with either invalid signature or failing `sample_tx:check/3` checks. That way, with current implementation, `apply_signed` returns an error and miner is never able to start mining, as creation of a block candidate fails!

I have added a check that newly created block for mining contains coinbase tx, and also added a comment in `aec_coinbase_tx` regarding adding transaction fees to miner's account.

lucafavatella commented on apps/aecore/src/txs/aec_coinbase_tx.erl:
```diff
@@ -28,6 +28,9 @@ check(#coinbase_tx{account = AccountPubkey}, Trees0, Height) ->
 
+%% Only aec_governance:block_mine_reward() is granted to miner's account here.
+%% Amount from all the fees of transactions included in the block
+%% is added to miner's account in aec_tx:apply_signed/3.
```
Thanks.

lucafavatella commented on apps/aecore/src/aec_mining.erl:
```diff
@@ -17,17 +17,18 @@
-            case aec_blocks:new(LastBlock, Txs, Trees) of
-                {ok, Block0} ->
+            Block0 = aec_blocks:new(LastBlock, Txs, Trees),
+            case aec_blocks:cointains_coinbase_tx(Block0) of
```
Thanks.

zp-sd commented on apps/aecore/src/aec_chain_state.erl:
```diff
@@ -527,6 +527,9 @@ update_top_block_hash(Hash, State) ->
 apply_node_transactions([Node|Left], Trees, State) ->
     Txs = aec_blocks:txs(Node#node.content),
     Height = node_height(Node),
+    %% TODO: verify that root hash of state tree after applying transactions
```
Well, actually, based on our `blocks.hrl`:
```
root_hash = <<0:?STATE_HASH_BYTES/unit:8>> :: state_hash(), % Hash of all state Merkle trees included in #block.trees
```
so we should check `root_hash` as a hash of all state trees.

Transactions' hash is stored in 
```
txs_hash = <<0:?TXS_HASH_BYTES/unit:8>> :: txs_hash(),
```
and this is already being verified as a part of block verification feature, when a block is received from the network.
